File: 2023/02/day02.py
def get_input(file):
    return quantize([line.strip()[5:].split(':') for line in file])

def part1(games):
    cubes = {'red': 12, 'green': 13, 'blue': 14}
    return sum([
        gid for gid, rounds in games.items()
        if all([is_possible(r, cubes) for r in rounds])
    ])

def part2(games):
    return sum([
        max([r.get('red', 0) for r in game]) *
        max([r.get('green', 0) for r in game]) *
        max([r.get('blue', 0) for r in game])
        for game in games.values()
    ])

def is_possible(rnd, cnts):
    return (rnd.get('red', 0) <= cnts['red'] and
            rnd.get('green', 0) <= cnts['green'] and
            rnd.get('blue', 0) <= cnts['blue'])

def quantize(games):
    return {
        int(g[0]): [
            {ps.split(' ')[-1]: int(ps.split(' ')[1]) for ps in r.split(',')}
            for r in g[-1].split(';')
        ] for g in games
    }

File: 2023/02/day02.ex
defmodule Day02 do
  def get_input(file) do
    file
    |> IO.read(:eof)
    |> String.split("\n", trim: true)
    |> Enum.map(fn line -> String.slice(line, 5..-1) |> String.split(":") end)
    |> quantize()
  end

  def part1(input) do
    cubes = %{"red" => 12, "green" => 13, "blue" => 14}
    for {id, game} <- input, Enum.all?(game, fn r -> possible?(r, cubes) end) do
      id
    end
    |> Enum.sum()
  end

  def part2(input) do
    for {_, game} <- input do
      Enum.max(for r <- game, do: Map.get(r, "red", 0)) *
      Enum.max(for r <- game, do: Map.get(r, "green", 0)) *
      Enum.max(for r <- game, do: Map.get(r, "blue", 0))
    end
    |> Enum.sum()
  end

  defp possible?(round, counts) do
    Map.get(round, "red", 0) <= counts["red"] and
    Map.get(round, "green", 0) <= counts["green"] and
    Map.get(round, "blue", 0) <= counts["blue"]
  end

  defp quantize(games) do
    for [id, g] <- games, into: %{} do
      {
        String.to_integer(id),
        for round <- String.split(g, ";") do
          for count <- String.split(round, ","), into: %{} do
            parts = String.split(count, " ", trim: true)
            {List.last(parts), List.first(parts) |> String.to_integer()}
          end
        end
      }
    end
  end
end
